File: benchmarks/livermore/k01_hydro.c
/* K1 — Hydro fragment (1-based indexing, matches Fortran exactly)
   Fortran: DIMENSION X(1001), Y(1001), Z(1012)
   DO K=1,N: X(K) = Q + Y(K)*(R*Z(K+10) + T*Z(K+11)) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     1001
#define NREPS 26147000

int main(void) {
    double x[1002], y[1002], z[1013];
    double spacer[40]; signel(spacer+1, 39);
    double q = spacer[28], r = spacer[30], t = spacer[36];

    signel(y+1, 1001);
    signel(z+1, 1012);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        for (long k = 1; k <= N; k++) {
            x[k] = q + y[k] * (r * z[k+10] + t * z[k+11]);
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("x[1] = %f\n", x[1]);
    return 0;
}

File: benchmarks/livermore/k01_hydro.f
C     Kernel 1 -- Hydro fragment
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C
      PROGRAM K01
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION X(1001), Y(1001), Z(1012)
      DIMENSION SPACER(39)
C
C     Initialize arrays using SIGNEL pattern
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      SCALED = 0.1D0
      DO 5 K = 1, 39
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          SPACER(K) = (BUZZ - FIZZ) * SCALED
    5 CONTINUE
      Q = SPACER(28)
      R = SPACER(30)
      T = SPACER(36)
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 10 K = 1, 1001
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          Y(K) = (BUZZ - FIZZ) * SCALED
   10 CONTINUE
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 15 K = 1, 1012
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          Z(K) = (BUZZ - FIZZ) * SCALED
   15 CONTINUE
C
C     Kernel loop
C
      N = 1001
      DO 100 REP = 1, 10000
          DO 90 K = 1, N
              X(K) = Q + Y(K) * (R * Z(K+10) + T * Z(K+11))
   90     CONTINUE
  100 CONTINUE
C
      WRITE(*,900) X(1)
  900 FORMAT('K1 hydro: x(1) = ', E25.15)
      END
